File: mantra_dev/mantra_dev/report/_qc_request_approval/qc_request_approval.js
// Copyright (c) 2024, Foram Shah and contributors
// For license information, please see license.txt

frappe.query_reports["QC Request Approval"] = {

	onload: function () {
		$(document).on("click", ".approvestockentry", function () {
			var button = $(this);

			var stockEntryId = button.data("stock_entry");
			var itemCode = button.data("item_code");
			var qty = button.data("qty");
			var purchaseReceipt = button.data("purchase_receipt");


			frappe.db.get_value("Item", itemCode, "has_serial_no", function (r) {
				let isSerialized = r && r.has_serial_no ? r.has_serial_no : 0;

				if (isSerialized) {
					frappe.prompt(
						[
							{
								label: "Quantity",
								fieldname: "qty",
								fieldtype: "Float",
								default: qty,
								read_only: 1,
							},
							{
								label: "Serial Numbers",
								fieldname: "serial_no",
								fieldtype: "Long Text",
								reqd: isSerialized ? 1 : 0,
								description: isSerialized ? "Enter one serial number per line." : "",
							}
						],
						(values) => {

							let serialNumbers = values.serial_no.split("\n").map(sn => sn.trim()).filter(sn => sn);

							// Check if serial numbers are unique
							let uniqueSerialNumbers = new Set(serialNumbers);

							if (uniqueSerialNumbers.size !== serialNumbers.length) {
								frappe.msgprint({
									title: __("Validation Error"),
									indicator: "red",
									message: __("Duplicate serial numbers are not allowed."),
								});
								return;
							}

							if (serialNumbers.length !== parseInt(qty)) {
								frappe.msgprint({
									title: __("Validation Error"),
									indicator: "red",
									message: __("The number of serial numbers (" + serialNumbers.length + ") must match the quantity (" + parseInt(qty) + ")."),
								});
								return;
							}

							frappe.call({

								method: "mantra_dev.backend_code.qc_module.get_valid_serial_numbers",
								args: {
									purchase_receipt_id: purchaseReceipt,
									item_code: itemCode,
									serial_numbers: values.serial_no ? values.serial_no.replace(/\n/g, ",") : "" 
								},

								callback: function (r) {

									if (r.message.status === "error") {
										frappe.msgprint({
											title: __("Validation Error"),
											indicator: "red",
											message: r.message.message,
										});
										return;
									}

									// If serial numbers are valid, add them to stock entry and approve
                                    frappe.call({
                                        method: "mantra_dev.backend_code.qc_module.add_serial_numbers_and_approve",
                                        args: {
                                            stock_entry: stockEntryId,
                                            item_code: itemCode,
                                            serial_numbers: values.serial_no.replace(/\n/g, ",")  // Convert newlines to commas
                                        },
                                        callback: function (res) {
                                            if (res.message.status === "success") {
                                                frappe.msgprint(res.message.message);
                                                frappe.query_report.refresh();

                                            } else {
                                                frappe.msgprint({
                                                    title: __("Error"),
                                                    indicator: "red",
                                                    message: res.message.message,
                                                });
                                            }
                                        }
                                    });
                                }
                            });
                        },

						"Enter Serial No for QC",
						"Proceed"
					);
				} else {
					// Approve non-serialized stock entry directly
					frappe.call({
						method: "mantra_dev.backend_code.qc_module.approve_stock_entry",
						args: {
							stock_entry: stockEntryId,
						},
						callback: function (r) {
							if (r.message.status === "success") {
								frappe.msgprint(r.message.message);
								frappe.query_report.refresh();

							} else {
								frappe.msgprint({
									title: __("Error"),
									indicator: "red",
									message: res.message.message,
								});
							}
						},
						error: function () {
							frappe.msgprint(__("An error occurred while approving the Stock Entry."));
						}
					});
				}
			});
		});
	},
};

File: mantra_dev/mantra_dev/report/_qc_request_approval/qc_request_approval.py
# Copyright (c) 2024, Foram Shah and contributors
# For license information, please see license.txt

import frappe
from frappe import _




def execute(filters=None):
  columns, data = get_columns(), get_data(filters)
  return columns, data




def get_columns():
  return [
      {"label": _("Purchase Receipt"), "fieldname": "VoucherNumber", "fieldtype": "Link", "options": "Purchase Receipt", "width": 200},
      {"label": _("Date"), "fieldname": "Date", "fieldtype": "Date", "width": 100},
      {"label": _("Item Code"), "fieldname": "ItemCode", "fieldtype": "Link", "options": "Item", "width": 100},
      {"label": _("Item Name"), "fieldname": "ItemName", "fieldtype": "Data", "width": 150},
      {"label": _("Accepted QTY"), "fieldname": "AcceptedQTY", "fieldtype": "Int", "width": 70},
      {"label": _("Warehouse"), "fieldname": "Warehouse", "fieldtype": "Data", "width": 150},
      {"label": _("Inspection Required"), "fieldname": "InspectionRequired", "fieldtype": "Check", "width": 70},
      {"label": _("Stock Entry"), "fieldname": "StockEntry", "fieldtype": "Link", "options": "Stock Entry", "width": 170},
      {"label": _("Approve"), "fieldname": "Approve", "fieldtype": "Button", "width": 90},
  
  ]

def get_data(filters):

    # Fetch the default inward warehouse from QC Settings
    inward_warehouse = frappe.db.get_value("QC Settings", "QC Settings", "default_inward_warehouse")

    if not inward_warehouse:
        frappe.throw(_("Default Inward Warehouse is not set in QC Settings."))

    data = frappe.db.sql("""
                       
    SELECT
        se.name AS StockEntry,
        sed.item_code AS ItemCode,
        sed.item_name AS ItemName,
        sed.qty AS AcceptedQTY,
        sed.t_warehouse AS Warehouse,
        pr.name AS VoucherNumber,
        pr.posting_date AS Date,
        sed.custom_inspection_required_before_transfer_warehouse AS InspectionRequired,
        CONCAT(
           '<button class="btn btn-primary pt-0 pb-0 approvestockentry" style="background-color: grey" ',
           'data-stock_entry="', se.name, '" ',
           'data-item_code="', sed.item_code, '" ',
           'data-qty="', sed.qty, '" ',
           'data-purchase_receipt="', pr.name, '" ',
           '>Approve</button>'
       ) AS Approve
    FROM
        `tabStock Entry` se
    JOIN
        `tabStock Entry Detail` sed ON sed.parent = se.name
    INNER JOIN
        `tabPurchase Receipt` pr ON pr.name = sed.reference_purchase_receipt
    WHERE
        se.docstatus = 0 
        AND
        sed.s_warehouse = %s
    ORDER BY
        pr.posting_date;
    
    
    """, (inward_warehouse,), as_dict=True)
    return data

